File: test/IECore/RunTimeTyped.py
##########################################################################
#
#  Copyright (c) 2007-2014, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

import sys
import unittest

import IECore

class TestRunTimeTyped( unittest.TestCase ) :

	def test( self ) :

		i = IECore.IntData( 10 )
		self.assertTrue( i.isInstanceOf( "IntData" ) )
		self.assertTrue( i.isInstanceOf( "Data" ) )
		self.assertTrue( i.isInstanceOf( "Object" ) )
		self.assertTrue( i.isInstanceOf( "RunTimeTyped" ) )

		self.assertTrue( i.isInstanceOf( IECore.TypeId.IntData ) )
		self.assertTrue( i.isInstanceOf( IECore.TypeId.Data ) )
		self.assertTrue( i.isInstanceOf( IECore.TypeId.Object ) )
		self.assertTrue( i.isInstanceOf( IECore.TypeId.RunTimeTyped ) )

		self.assertEqual( i.baseTypeId(), IECore.TypeId.Data )

		self.assertTrue( IECore.IntData.inheritsFrom( "Data" ) )
		self.assertTrue( IECore.IntData.inheritsFrom( "Object" ) )
		self.assertTrue( IECore.IntData.inheritsFrom( "RunTimeTyped" ) )

		self.assertTrue( IECore.IntData.inheritsFrom( IECore.TypeId.Data ) )
		self.assertTrue( IECore.IntData.inheritsFrom( IECore.TypeId.Object ) )
		self.assertTrue( IECore.IntData.inheritsFrom( IECore.TypeId.RunTimeTyped ) )

		self.assertTrue( IECore.Data.inheritsFrom( IECore.TypeId.Object ) )
		self.assertTrue( IECore.Data.inheritsFrom( IECore.TypeId.RunTimeTyped ) )

		self.assertTrue( IECore.Object.inheritsFrom( IECore.TypeId.RunTimeTyped ) )

		self.assertEqual( IECore.RunTimeTyped.typeNameFromTypeId( IECore.TypeId.IntData ), "IntData" )
		self.assertEqual( IECore.RunTimeTyped.typeIdFromTypeName( "IntData" ), IECore.TypeId.IntData )

	def testStaticTypeBindings( self ) :

		import IECore

		typeNames={}
		typeIds = {}

		# Make that all classes derived from RunTimeTyped correctly bind staticTypeName and staticTypeId
		for typeName in dir(IECore):

			t = getattr(IECore, typeName)

			baseClasses = []

			# Iteratively expand base classes all the way to the top
			if hasattr(t, "__bases__"):
				baseClasses = list( t.__bases__ )

				i=0
				while i < len(baseClasses):
					for x in baseClasses[i].__bases__:
						baseClasses.extend([x])

					i = i + 1

			if IECore.RunTimeTyped in baseClasses:

				baseIds = IECore.RunTimeTyped.baseTypeIds( t.staticTypeId() )
				baseIds = set( [ int(x) for x in baseIds ] )

				self.assertTrue( hasattr(t, "staticTypeName") )
				self.assertTrue( hasattr(t, "staticTypeId") )
				self.assertTrue( hasattr(t, "baseTypeId") )
				self.assertTrue( hasattr(t, "baseTypeName") )

				self.assertTrue( t.staticTypeId() in IECore.TypeId.values.keys() )

				# Make sure that no 2 IECore classes provide the same typeId or typeName
				if t.staticTypeName() in typeNames:
					raise RuntimeError( "'%s' does not have a unique RunTimeTyped static type name (conflicts with '%s')." % ( t.__name__ , typeNames[t.staticTypeName()] ))

				if t.staticTypeId() in typeIds:
					raise RuntimeError( "'%s' does not have a unique RunTimeTyped static type id (conflicts with '%s')." % (t.__name__ , typeIds[t.staticTypeId()] ))

				self.assertEqual( IECore.RunTimeTyped.typeNameFromTypeId( t.staticTypeId() ), t.staticTypeName() )
				self.assertEqual( IECore.RunTimeTyped.typeIdFromTypeName( t.staticTypeName() ), t.staticTypeId() )

				for base in baseClasses :

					if issubclass( base, IECore.RunTimeTyped ) :

						self.assertNotEqual( base.staticTypeId(), t.staticTypeId() )
						if not base.staticTypeId() in IECore.RunTimeTyped.baseTypeIds( t.staticTypeId() ):
							raise Exception( "'%s' does not have '%s' in its RunTimeTyped base classes, even though Python says it derives from it." % (t.staticTypeName(), base.staticTypeName() ))

				typeNames[t.staticTypeName()] = t.__name__
				typeIds[t.staticTypeId()] = t.__name__


	def testRegisterRunTimeTyped( self ) :

		# should raise because given type ID is different than the FileSequenceParameter type id
		self.assertRaises( Exception, IECore.registerRunTimeTyped, IECore.FileSequenceParameter, 100009 )
		# should raise because SequenceLsOp is registered with dynamic type id.
		self.assertRaises( Exception, IECore.registerRunTimeTyped, IECore.SequenceLsOp, 100009 )
		# should raise because FileSequenceParameter is registered with a non-dynamic type id
		self.assertRaises( Exception, IECore.registerRunTimeTyped, IECore.FileSequenceParameter )

		# should not raise because SequenceLsOp was already registered with a dynamic type id
		IECore.registerRunTimeTyped( IECore.SequenceLsOp )

		self.assertEqual( IECore.TypeId.OptionalCompoundParameter, IECore.OptionalCompoundParameter.staticTypeId() )
		self.assertEqual( IECore.TypeId.OptionalCompoundParameter, IECore.OptionalCompoundParameter( "", "" ).typeId() )

		self.assertTrue( IECore.OptionalCompoundParameter.inheritsFrom( "CompoundParameter" ) )
		self.assertTrue( IECore.OptionalCompoundParameter.inheritsFrom( IECore.TypeId.CompoundParameter ) )
		self.assertTrue( IECore.OptionalCompoundParameter.inheritsFrom( "Parameter" ) )
		self.assertTrue( IECore.OptionalCompoundParameter.inheritsFrom( IECore.TypeId.Parameter ) )
		self.assertTrue( IECore.OptionalCompoundParameter.inheritsFrom( "RunTimeTyped" ) )
		self.assertTrue( IECore.OptionalCompoundParameter.inheritsFrom( IECore.TypeId.RunTimeTyped ) )

		self.assertTrue( IECore.OptionalCompoundParameter( "", "" ).isInstanceOf( "OptionalCompoundParameter" ) )
		self.assertTrue( IECore.OptionalCompoundParameter( "", "" ).isInstanceOf( IECore.TypeId.OptionalCompoundParameter ) )
		self.assertTrue( IECore.OptionalCompoundParameter( "", "" ).isInstanceOf( "CompoundParameter" ) )
		self.assertTrue( IECore.OptionalCompoundParameter( "", "" ).isInstanceOf( IECore.TypeId.CompoundParameter ) )
		self.assertRaises( TypeError, IECore.OptionalCompoundParameter( "", "" ).isInstanceOf, 10 )

	def testTypeNameFromRunTimeTypedTypeId( self ) :

		self.assertEqual( IECore.RunTimeTyped.typeIdFromTypeName( "RunTimeTyped" ), IECore.TypeId.RunTimeTyped )

	def testRunTimeTypedTypeIdFromTypeName( self ) :

		self.assertEqual( IECore.RunTimeTyped.typeNameFromTypeId( IECore.TypeId.RunTimeTyped ), "RunTimeTyped" )

	def testInheritsFromWithTwoArguments( self ) :

		self.assertTrue( IECore.RunTimeTyped.inheritsFrom( IECore.TypeId.Object, IECore.TypeId.RunTimeTyped ) )
		self.assertTrue( IECore.RunTimeTyped.inheritsFrom( "Object", "RunTimeTyped" ) )

		self.assertTrue( IECore.RunTimeTyped.inheritsFrom( IECore.TypeId.CompoundObject, IECore.TypeId.Object ) )
		self.assertTrue( IECore.RunTimeTyped.inheritsFrom( "CompoundObject", "Object" ) )

		self.assertFalse( IECore.RunTimeTyped.inheritsFrom( IECore.TypeId.CompoundObject, IECore.TypeId.Writer ) )
		self.assertFalse( IECore.RunTimeTyped.inheritsFrom( "CompoundObject", "Writer" ) )

	def testRegisterPrefixedTypeName( self ) :

		class Prefixed( IECore.Op ) :

			def __init__( self ) :

				IECore.Op.__init__( self, "", IECore.IntParameter( "result", "" ) )

		prefixedTypeName = "SomeModuleName::Prefixed"
		IECore.registerRunTimeTyped( Prefixed, typeName = prefixedTypeName )

		self.assertEqual( Prefixed.staticTypeName(), prefixedTypeName )
		self.assertEqual( IECore.RunTimeTyped.typeIdFromTypeName( Prefixed.staticTypeName() ), Prefixed.staticTypeId() )

		p = Prefixed()
		self.assertEqual( p.typeName(), prefixedTypeName )
		self.assertEqual( p.typeId(), IECore.RunTimeTyped.typeIdFromTypeName( Prefixed.staticTypeName() ) )

		self.assertTrue( p.isInstanceOf( IECore.Op.staticTypeId() ) )
		self.assertTrue( p.isInstanceOf( IECore.Parameterised.staticTypeId() ) )
		self.assertTrue( p.isInstanceOf( IECore.RunTimeTyped.staticTypeId() ) )

	def testClassInPlaceOfTypeId( self ) :

		# check that we can pass the python class itself
		# where C++ would like a TypeId.
		self.assertTrue( IECore.Data.inheritsFrom( IECore.RunTimeTyped ) )
		self.assertTrue( IECore.Data.inheritsFrom( IECore.Object ) )
		self.assertFalse( IECore.Data.inheritsFrom( IECore.Parameter ) )
		self.assertRaises( Exception, IECore.Data.inheritsFrom, dict )

		# check that the converter mechanism doesn't both the
		# reference count for the class objects
		r = sys.getrefcount( IECore.Data )
		IECore.IntData.inheritsFrom( IECore.Data )
		self.assertEqual( r, sys.getrefcount( IECore.Data ) )

if __name__ == "__main__":
    unittest.main()

File: src/IECore/RunTimeTyped.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007-2009, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include "IECore/RunTimeTyped.h"

using namespace IECore;

RunTimeTyped::RunTimeTyped()
{
}

RunTimeTyped::~RunTimeTyped()
{
}

TypeId RunTimeTyped::typeId() const
{
	return staticTypeId();
}

std::string RunTimeTyped::typeName() const
{
	return staticTypeName();
}

TypeId RunTimeTyped::staticTypeId()
{
	return RunTimeTypedTypeId;
}

std::string RunTimeTyped::staticTypeName()
{
	return "RunTimeTyped";
}

TypeId RunTimeTyped::baseTypeId()
{
	return InvalidTypeId;
}
		
std::string RunTimeTyped::baseTypeName()
{
	return "InvalidType";
}	

bool RunTimeTyped::isInstanceOf( TypeId typeId ) const
{
	return typeId==staticTypeId();
}

bool RunTimeTyped::isInstanceOf( const std::string &typeName ) const
{
	return typeName==staticTypeName();
}

bool RunTimeTyped::inheritsFrom( TypeId typeId )
{
	return false;
}

bool RunTimeTyped::inheritsFrom( const std::string &typeName )
{
	return false;
}

void RunTimeTyped::registerType( TypeId derivedTypeId, const std::string &derivedTypeName, TypeId baseTypeId )
{
	BaseTypeRegistryMap &baseRegistry = baseTypeRegistry();
#ifndef NDEBUG
	{
		BaseTypeRegistryMap::const_iterator it = baseRegistry.find( derivedTypeId );
		if ( it != baseRegistry.end() )
		{
			assert( it->second == baseTypeId );
		}	
	}
#endif	
	
	baseRegistry[ derivedTypeId ] = baseTypeId;
	
	DerivedTypesRegistryMap &derivedRegistry = derivedTypesRegistry();
	derivedRegistry[ baseTypeId ].insert( derivedTypeId );
	
	/// Put in id->name map
	TypeIdsToTypeNamesMap &idsToNames = typeIdsToTypeNames();
#ifndef NDEBUG
	{
		TypeIdsToTypeNamesMap::const_iterator it = idsToNames.find( derivedTypeId );
		if ( it != idsToNames.end() )
		{
			assert( it->second == derivedTypeName );
		}
	}	
#endif
	idsToNames[ derivedTypeId ] = derivedTypeName;
	
	/// Put in name->id map
	TypeNamesToTypeIdsMap &namesToIds = typeNamesToTypeIds();
#ifndef NDEBUG
	{
		TypeNamesToTypeIdsMap::const_iterator it = namesToIds.find( derivedTypeName );
		if ( it != namesToIds.end() )
		{
			assert( it->second == derivedTypeId );
		}	
	}
#endif
	namesToIds[ derivedTypeName ] = derivedTypeId;
	
}

RunTimeTyped::BaseTypeRegistryMap &RunTimeTyped::baseTypeRegistry()
{
	static BaseTypeRegistryMap *registry = new BaseTypeRegistryMap();
	
	assert( registry );
	return *registry;
}

RunTimeTyped::DerivedTypesRegistryMap &RunTimeTyped::derivedTypesRegistry()
{
	static DerivedTypesRegistryMap *registry = new DerivedTypesRegistryMap();
	
	assert( registry );	
	return *registry;
}

TypeId RunTimeTyped::baseTypeId( TypeId typeId )
{
	BaseTypeRegistryMap &baseRegistry = baseTypeRegistry();
	BaseTypeRegistryMap::const_iterator it = baseRegistry.find( typeId );
	
	if ( it == baseRegistry.end() )
	{
		return InvalidTypeId;
	}
	else
	{
		return it->second;
	}
}
				
const std::vector<TypeId> &RunTimeTyped::baseTypeIds( TypeId typeId )
{
	BaseTypesRegistryMap &baseTypes = completeBaseTypesRegistry();
	
	BaseTypesRegistryMap::iterator it = baseTypes.find( typeId );
	if ( it != baseTypes.end() )
	{
		return it->second;
	}
	
	baseTypes.insert( BaseTypesRegistryMap::value_type( typeId, std::vector<TypeId>() ) );
	it = baseTypes.find( typeId );
	assert( it != baseTypes.end() );
	
	TypeId baseType = baseTypeId( typeId );
	while ( baseType != InvalidTypeId )
	{
		it->second.push_back( baseType );
		baseType = baseTypeId( baseType );
	}
	
	return it->second;
}
		
const std::set<TypeId> &RunTimeTyped::derivedTypeIds( TypeId typeId )
{	
	DerivedTypesRegistryMap &derivedTypes = completeDerivedTypesRegistry();	
	DerivedTypesRegistryMap::iterator it = derivedTypes.find( typeId );
	
	if ( it == derivedTypes.end() )
	{
		derivedTypes.insert( DerivedTypesRegistryMap::value_type( typeId, std::set<TypeId>() ) );
		it = derivedTypes.find( typeId );
		assert( it != derivedTypes.end() );
		
		// Walk over the hierarchy of derived types
		derivedTypeIdsWalk( typeId, it->second );
	}
	
	return it->second;	
}

void RunTimeTyped::derivedTypeIdsWalk( TypeId typeId, std::set<TypeId> &typeIds )
{
	DerivedTypesRegistryMap &derivedRegistry = derivedTypesRegistry();	
	DerivedTypesRegistryMap::const_iterator it = derivedRegistry.find( typeId );
	if ( it == derivedRegistry.end() )
	{
		/// Termination condition: No derived types
		return;
	}
	
	for ( std::set<TypeId>::const_iterator typesIt = it->second.begin(); typesIt != it->second.end(); ++ typesIt )
	{
		typeIds.insert( *typesIt );
		
		// Recurse down into derived types	
		derivedTypeIdsWalk( *typesIt, typeIds );
	}
}

RunTimeTyped::BaseTypesRegistryMap &RunTimeTyped::completeBaseTypesRegistry()
{
	static BaseTypesRegistryMap *baseTypes = new BaseTypesRegistryMap();
	assert( baseTypes );
	return *baseTypes;
}

RunTimeTyped::DerivedTypesRegistryMap &RunTimeTyped::completeDerivedTypesRegistry()
{
	static DerivedTypesRegistryMap *derivedTypes = new DerivedTypesRegistryMap();
	assert( derivedTypes );
	return *derivedTypes;
}

TypeId RunTimeTyped::typeIdFromTypeName( const std::string &typeName )
{
	TypeNamesToTypeIdsMap &namesToIds = typeNamesToTypeIds();
	TypeNamesToTypeIdsMap::const_iterator it = namesToIds.find( typeName );
	if( it == namesToIds.end() )
	{
		return InvalidTypeId;
	}
	return it->second;
}

std::string RunTimeTyped::typeNameFromTypeId( TypeId typeId )
{
	TypeIdsToTypeNamesMap &idsToNames = typeIdsToTypeNames();
	TypeIdsToTypeNamesMap::const_iterator it = idsToNames.find( typeId );
	if( it == idsToNames.end() )
	{
		return "";
	}
	return it->second;
}

RunTimeTyped::TypeIdsToTypeNamesMap &RunTimeTyped::typeIdsToTypeNames()
{
	static TypeIdsToTypeNamesMap *m = new TypeIdsToTypeNamesMap();
	assert( m );
	return *m;
}

RunTimeTyped::TypeNamesToTypeIdsMap &RunTimeTyped::typeNamesToTypeIds()
{
	static TypeNamesToTypeIdsMap *m = new TypeNamesToTypeIdsMap();
	assert( m );
	return *m;
}
